### ISSUE TITLE
Duda sobre crazy-seqs... ay! no! que diga lazy-seqs!

### ISSUE THREAD

**Issue by mi-mina** [Grupete/jau-tu#2]
A ver... mientras la cosa se mueve y no, yo en mis ratos libres (que son minúsculos ahora mismo) tratando de avanzar un poco con el clojure y... tratando de hacer el ejercicio 118 de 4clojure, Re-implement map, mi primer intento ha sido:
`(defn _map  [f x] (reduce #(conj %1 (f %2)) [] x))`,
que funciona guapi y yo tó contenta porque me ha salido idiomático casi del tirón (después de un intento con `loop`, que he dicho, seguro que a clojure le gusta más de otra manera) peeeero no me devuelve una lazy-seq, por lo que para seqs infinitas, como esta `(range)` el ordenador se queda locor :sweat_smile:

Entonces digo, pues lazy-seq al canto! así:
`(defn _map  [f x] (lazy-seq (reduce #(conj %1 (f %2)) [] x)))`, 
pero el Light Table me dice, jar! que te crees tú eso y me suelta un java.lang.OutOfMemoryError

Luego, haciendo trampas sin querer, veo por ahí la definición de map, que usa lazy-seq, pero con una función recursiva, tal que así:
`(defn map
  ([f x]
   (lazy-seq
    (when-let [s (seq x)]
      (cons (f (first s)) (map f (rest s)))))))`

Y digo yo ¿Es por culpa de `reduce` que no me funciona a mí?

**Comment by tangrammer** [comment]
Yes, reduce is not lazy! por eso que después el amigo Rich Hickey tuvo que añadir [reductions](https://clojuredocs.org/clojure.core/reductions) que por lo visto andan relacionadas con [transducers](http://clojure.org/transducers) (esto todavía no lo leí ... me estoy controlando :) )

Si quieres puedo publicar mi solución vale?

**Comment by tangrammer** [comment]
aquí una buena pagina para empezar con transducers
http://stackoverflow.com/questions/26317325/can-someone-explain-clojure-transducers-to-me-in-simple-terms

**Comment by mi-mina** [comment]
Gracias!!! O sea que si quiero que algo sea lazy no puedo usar reduce.... Le estuve echando un vistazo a reductions pero me dio miedito, jajajajaj... y no llegué a conseguir que funcionara, pero vi que alguien lo uso para la solución de este problema

**Comment by tangrammer** [comment]
:) de hecho reduce no devuelve una seq por defecto sino el resultado (a lo
mejor un numero) de:

```

f should be a function of 2 arguments. If val is not supplied,
returns the result of applying f to the first 2 items in coll, then
applying f to that result and the 3rd item, etc. If coll contains no
items, f must accept no arguments as well, and reduce returns the
result of calling f with no arguments.  If coll has only 1 item, it
is returned and f is not called.  If val is supplied, returns the
result of applying f to val and the first item in coll, then
applying f to that result and the 2nd item, etc. If coll contains no
items, returns val and f is not called.

```

**Comment by mi-mina** [comment]
hostia claro! y para devolver ese resultado tiene que recorrer todos los argumentos...

**Comment by tangrammer** [comment]
yeah! de todas maneras yo nunca he usado la function lazy-seq todavía :) .
Aunque he visto casos que queda muy chula

2015-04-26 21:33 GMT+02:00 Esperanza Moreno Cruz notifications@github.com:
